### PR TITLE
Make OtherProtocol behave like RawAnnotation

### PR DESCRIPTION
**Pull request by chiroptical**
This is a very similar change to the Kotlin `Annotation/Interface`. We are circumventing the error:

```
test/Spec.hs:19:1: error:
    Data constructor not in scope: CaseIterable :: Protocol
```

Also discovered some issues related to `dataRawValue`s and empty `Protocol/Annotation`

### CODE REVIEW

File: src/Moat.hs [hunk 1]
```diff
@@ -1734,20 +1736,16 @@ giveBase r ps = \case
...
-      Equatable -> "Equatable"
-      Hashable -> "Hashable"
-      Codable -> "Codable"
-      OtherProtocol p -> p
+applyBase :: (Bool, Maybe MoatType, [Protocol]) -> Exp -> Q Exp
+applyBase (b, r, ps) (ParensE -> s) = do
+  protoAnnotations <- protosExp ps
+  pure $
+    if b
+      then AppE (AppE (AppE (VarE 'giveBase) (rawValueE r)) protoAnnotations) s
```
Comment (parsonsmatt) [inline src/Moat.hs:1744]: so I generally either prefer to use `AppE` infix to make application more obvious, or, if we're in `Q`, we can use the quasi quoter.

```haskell
-- with infix AppE
pure $ VarE 'giveBase `AppE` rawValueE r `AppE` protoAnnotations `AppE` s

-- factor it out with `foldl'`:
pure $ foldl' AppE (VarE 'giveBase) [rawValueE r, protoAnnotations, s]

-- or, with quasiquoter,
[| giveBase $(pure (rawValueE r)) $(lift ps) $(pure s) |]
```

A QuasiQuote has the following syntax:

```
[parser| some string |]
```

If you leave out the `parser`, then it uses the default Haskell parser `e`. `[d| ... |]` parses Haskell declarations and `[t| ... |]` parses Haskell types. So `[t| Maybe Int |] :: Q Type`, and `[e|Just 10|] :: Q Exp` and `[d|class Foo a where foo :: Int |] :: Q [Dec]`.

Then we've got the splice syntax - `$( something )`. The type of `something` depends on where you use it - in the Haskell quoter, it expects a `Q a` value. That's why we need to write `pure` up there - to take it from `Exp -> Q Exp`.

File: src/Moat.hs [hunk 2]
```diff
@@ -1734,20 +1736,16 @@ giveBase r ps = \case
...
+      else s
+
+protosExp :: [Protocol] -> Q Exp
+protosExp = fmap ListE <$> traverse Syntax.lift
```
Comment (parsonsmatt) [inline src/Moat.hs:1748]: i think this can be simply `lift`

**Review by parsonsmatt — APPROVED** [review]
looks great! just a style note

File: src/Moat.hs (hunk 2 above)
Comment (chiroptical) [reply src/Moat.hs]: Addressed here: 117d8c7

File: src/Moat.hs (hunk 1 above)
Comment (chiroptical) [reply src/Moat.hs]: Addressed here: 117d8c7

**Review by parsonsmatt — APPROVED** [review]

**Pull request merged**